lib/linkerd2-trace-context/src/layer.rs

```diff
@@ -0,0 +1,184 @@
+/// present, a new span will be started in the current trace by creating a new
+/// random span id setting it into the `traceparent` header before forwarding
+/// the request.  If the sampled bit of the header was set, we emit metadata
+/// about the span to the given SpanSink when the span is complete, i.e. when
+/// we receive the response.
+pub fn layer<Sink>(sink: Sink) -> Layer<Sink> {
+    Layer { sink }
+}
+
+// === impl Layer ===
+
+impl<Make, Sink> tower::layer::Layer<Make> for Layer<Sink>
+where
+    Sink: Clone,
+{
+    type Service = Stack<Make, Sink>;
+
+    fn layer(&self, inner: Make) -> Self::Service {
+        Stack {
+            inner,
+            sink: self.sink.clone(),
+        }
+    }
+}
+
+// === impl Stack ===
+
+impl<Target, Make, Sink> tower::Service<Target> for Stack<Make, Sink>
+where
+    Make: tower::Service<Target>,
+    Sink: Clone,
+{
+    type Response = Service<Make::Response, Sink>;
+    type Error = Make::Error;
+    type Future = MakeFuture<Make::Future, Sink>;
+
+    fn poll_ready(&mut self) -> Poll<(), Make::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: Target) -> Self::Future {
+        let inner = self.inner.call(target);
+
+        MakeFuture {
+            inner,
+            sink: Some(self.sink.clone()),
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<Fut: Future, Sink> Future for MakeFuture<Fut, Sink> {
+    type Item = Service<Fut::Item, Sink>;
+    type Error = Fut::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let sink = self.sink.take().expect("poll called after ready");
+        Ok(Async::Ready(Service { inner, sink }))
+    }
+}
+
+// === impl Service ===
+
+impl<Svc, Body, Sink> tower::Service<http::Request<Body>> for Service<Svc, Sink>
+where
+    Svc: tower::Service<http::Request<Body>>,
+    Sink: SpanSink + Clone,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = futures::future::Either<Svc::Future, SpanFuture<Svc::Future, Sink>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Svc::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: http::Request<Body>) -> Self::Future {
+        let mut trace_context = propagation::unpack_trace_context(&request);
+        let mut path = None;
+
+        if let Some(ref mut context) = trace_context {
+            trace!(message = "got trace context", ?context);
+            propagation::increment_span_id(&mut request, context);
+            // If we plan to sample this span, we need to copy the path from
+            // the request before dispatching it to inner.
+            if context.is_sampled() {
+                path = Some(request.uri().path().to_string());
+            }
+        }
+
+        let f = self.inner.call(request);
+
+        if let Some(propagation::TraceContext {
+            propagation: _,
+            version: _,
```

Comment:
these are unused? you can elide these and include, `..` at the end (to account for any/all unknown fields)